src/Microsoft.Extensions.Hosting.Abstractions/IHostLifetime.cs

```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Hosting
+{
+    public interface IHostLifetime
```

Comment:
These method names need work. OnStarted blocks the start process, so it's more like OnStarting.

OnStopping initiates shutdown indirectly, chaining things like CTL+C to `IApplicationLifetime.StopApplication()`. It's not clear why it needs a callback do to that, it could Inject IApplicationLifetime directly.

StopAsync informs the HostLifetime instance that the host was stopped. This seems more like Dispose.

Reply:
> StopAsync informs the HostLifetime instance that the host was stopped. This seems more like Dispose.

Lets not conflate StopAsync and Dispose again 😄 . It'll already automagically get disposed if it implements IDisposable.

Reply:
There are weirdnesses here that I want to understand. Some of which might be a bit existential.

Why does IHostLifetime need to notify IApplicationLifetime if it is called by the IHost and the IHost already has IApplicationLifetime?

I don't understand why OnStarted isn't just `public Task OnStarted` where the `IHost` will await it, allowing the specific HostLifetime to delay if it wants to or just return. Doing this would what I think is the contract clearer.

The delegate appears to only be useful in a Host that doesn't use `IApplicationLifetime`. In our host implementations it is never going to be different to what is currently passed to the `HostLifetime`, correct? If this is true is not using `IApplicationLifetime` something we would expect an `IHost` to do, or do we view `IApplicationLifetime` as an integral part of the contract that a Host should adhere to?

Reply:
Agreed, I don't see why the host needs to register with StopAsync if all it's going to register is a call to IApplicationLifetime.StopApplication. The IHostLifetime can call IApplicationLifetime directly.

And yes, OnStarted could go with a task model, it would just have to be cancellable.

@davidfowl ?

Reply:
> I don't understand why OnStarted isn't just public Task OnStarted where the IHost will await it, allowing the specific HostLifetime to delay if it wants to or just return. Doing this would what I think is the contract clearer.

That works. IMO it's a calling pattern. We're talking about calling a method when something starts vs waiting until something starts to complete the Task returned. Semantically, calling StartAsync on IHostLifetime is bizarre as you're really asking the object to call you back when it has started. It's more like `IHostLifetime.WaitForStartAsync()`. See our HttpResponse. OnStarting event for a similar pattern.

> The delegate appears to only be useful in a Host that doesn't use IApplicationLifetime. In our host implementations it is never going to be different to what is currently passed to the HostLifetime, correct? If this is true is not using IApplicationLifetime something we would expect an IHost to do, or do we view IApplicationLifetime as an integral part of the contract that a Host should adhere to?
  
> Agreed, I don't see why the host needs to register with StopAsync if all it's going to register is a call to IApplicationLifetime.StopApplication. The IHostLifetime can call IApplicationLifetime directly.

I feel like I'm missing something here. The entire point of this interface is the abstract the lifetime from the IHost implementation so that things like Run and RunAsync can be implemented without hardcoding ties into various lifetime mechanisms. The IApplicationLifetime is about firing host events as they happen so the application can react.

Reply:
I was thinking of a more succinct way to put this:

- The IHostLifetime is used by the host. After a bit of thinking, I think they should potentially stack (support multiple implementations).
- The IApplicationLifetime is consumed by the application and implemented by the host.